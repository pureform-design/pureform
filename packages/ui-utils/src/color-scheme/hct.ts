export { Hct } from "@material/material-color-utilities";
