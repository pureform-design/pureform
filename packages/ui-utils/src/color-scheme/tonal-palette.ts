import {
    argbFromHex,
    type Hct,
    hexFromArgb,
    TonalPalette as MaterialTonalPalette,
} from "@material/material-color-utilities";
import type { HexColor } from "../types";

export class TonalPalette {
    private _matTp: MaterialTonalPalette;

    private constructor(tp: MaterialTonalPalette);
    private constructor(hax: HexColor);
    private constructor(hex: HexColor | MaterialTonalPalette) {
        if (typeof hex === "string") {
            this._matTp = MaterialTonalPalette.fromInt(argbFromHex(hex));
        } else {
            this._matTp = hex;
        }
    }

    public static fromHex(hex: HexColor) {
        return new TonalPalette(hex);
    }

    public static fromArgb(argb: number) {
        return new TonalPalette(hexFromArgb(argb) as HexColor);
    }

    public static fromHueAndChroma(hue: number, chroma: number) {
        const tp = MaterialTonalPalette.fromHueAndChroma(hue, chroma);
        return new TonalPalette(tp);
    }

    private static throwWhenInvalidTone(tone: number) {
        if (tone < 0 || tone > 100) {
            throw new Error(`Invalid tone: ${tone}, must be between 0 and 100`);
        }
    }

    public toneHex(tone: number): HexColor {
        TonalPalette.throwWhenInvalidTone(tone);

        return hexFromArgb(this._matTp.tone(tone)) as HexColor;
    }

    public toneHct(tone: number): Hct {
        TonalPalette.throwWhenInvalidTone(tone);

        return this._matTp.getHct(tone);
    }

    public toneArgb(tone: number): number {
        TonalPalette.throwWhenInvalidTone(tone);

        return this._matTp.tone(tone);
    }
}
