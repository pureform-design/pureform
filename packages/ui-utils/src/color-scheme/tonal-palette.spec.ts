import { TonalPalette } from "./tonal-palette";
import { Hct } from "./hct";
import { describe, expect, it } from "vitest";

const GREEN_ARGB = 0xff00ff00;

describe("TonalPalette", () => {
    it("should be able to create a tonal palette instance", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        expect(tp).toBeInstanceOf(TonalPalette);
    });

    it("should be able to get a hex color from tone 0 to 100", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        for (let i = 0; i <= 100; i++) {
            const hex = tp.toneHex(i);
            expect(hex).toContain("#");
        }
    });

    it("should be able to get ARGB int color from tone 0 to 100", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        for (let i = 0; i <= 100; i++) {
            const argb = tp.toneArgb(i);
            expect(argb).toBeTypeOf("number");
        }
    });

    it("should be able to get HCT color from tone 0 to 100", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        for (let i = 0; i <= 100; i++) {
            const hct = tp.toneHct(i);
            expect(hct).toBeInstanceOf(Hct);
        }
    });

    it("should match snapshots hex colors from tone 0 to 100", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);
        for (let i = 0; i <= 100; i++) {
            const hex = tp.toneHex(i);
            expect(hex).toMatchSnapshot();
        }
    });

    it("should throw when trying to get a hex color from a tone outside of the range", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        expect(() => tp.toneHex(-1)).toThrow();
        expect(() => tp.toneHex(101)).toThrow();
    });

    it("should throw when trying to get an ARGB int color from a tone outside of the range", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        expect(() => tp.toneArgb(-1)).toThrow();
        expect(() => tp.toneArgb(101)).toThrow();
    });

    it("should throw when trying to get an HCT color from a tone outside of the range", () => {
        const tp = TonalPalette.fromArgb(GREEN_ARGB);

        expect(() => tp.toneHct(-1)).toThrow();
        expect(() => tp.toneHct(101)).toThrow();
    });
});
