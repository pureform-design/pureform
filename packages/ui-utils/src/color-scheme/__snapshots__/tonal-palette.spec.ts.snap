// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 1`] = `"#000000"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 2`] = `"#000500"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 3`] = `"#000a00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 4`] = `"#000f00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 5`] = `"#001200"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 6`] = `"#001500"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 7`] = `"#001800"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 8`] = `"#001b00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 9`] = `"#001d00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 10`] = `"#001f00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 11`] = `"#002200"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 12`] = `"#002400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 13`] = `"#002600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 14`] = `"#002900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 15`] = `"#002b00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 16`] = `"#002d00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 17`] = `"#013000"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 18`] = `"#013200"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 19`] = `"#013500"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 20`] = `"#013700"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 21`] = `"#013a00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 22`] = `"#013c00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 23`] = `"#013f00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 24`] = `"#014100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 25`] = `"#014400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 26`] = `"#014600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 27`] = `"#014900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 28`] = `"#014b00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 29`] = `"#014e00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 30`] = `"#015100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 31`] = `"#015300"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 32`] = `"#015600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 33`] = `"#015900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 34`] = `"#015b00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 35`] = `"#025e00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 36`] = `"#026100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 37`] = `"#026300"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 38`] = `"#026600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 39`] = `"#026900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 40`] = `"#026c00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 41`] = `"#026e00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 42`] = `"#027100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 43`] = `"#027400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 44`] = `"#027700"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 45`] = `"#027a00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 46`] = `"#027c00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 47`] = `"#027f00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 48`] = `"#028200"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 49`] = `"#028500"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 50`] = `"#028800"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 51`] = `"#038b00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 52`] = `"#038e00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 53`] = `"#039100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 54`] = `"#039400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 55`] = `"#039600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 56`] = `"#039900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 57`] = `"#039c00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 58`] = `"#039f00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 59`] = `"#03a200"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 60`] = `"#03a500"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 61`] = `"#03a800"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 62`] = `"#03ab00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 63`] = `"#03ae00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 64`] = `"#03b100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 65`] = `"#03b400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 66`] = `"#03b700"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 67`] = `"#03ba00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 68`] = `"#03be00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 69`] = `"#03c100"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 70`] = `"#03c400"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 71`] = `"#03c700"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 72`] = `"#03ca00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 73`] = `"#03cd00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 74`] = `"#03d000"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 75`] = `"#02d300"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 76`] = `"#02d600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 77`] = `"#02d900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 78`] = `"#02dd00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 79`] = `"#02e000"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 80`] = `"#02e300"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 81`] = `"#02e600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 82`] = `"#02e900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 83`] = `"#01ed00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 84`] = `"#01f000"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 85`] = `"#01f300"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 86`] = `"#01f600"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 87`] = `"#00f900"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 88`] = `"#00fd00"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 89`] = `"#28ff1d"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 90`] = `"#5aff47"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 91`] = `"#77ff61"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 92`] = `"#8dff76"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 93`] = `"#9fff88"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 94`] = `"#afff99"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 95`] = `"#bdffa9"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 96`] = `"#cbffb8"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 97`] = `"#d7ffc6"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 98`] = `"#e2ffd4"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 99`] = `"#edffe1"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 100`] = `"#f7ffee"`;

exports[`TonalPalette > should match snapshots hex colors from tone 0 to 100 101`] = `"#ffffff"`;
